NIFI-9810 Isolate and deprecate RocksDB repository

kevdoran opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-9810](https://issues.apache.org/jira/browse/NIFI-9810) This PR extracts the RocksDBFlowFileRepository implementation to its own ~maven module~ NAR and excludes it from the assembly convenience binary. It is also marked as deprecated, and planned to be dropped as part of NiFi 2.0.

- Bump version to 6.29.5 for arm64 compatibilty
- Extract RocksDBFlowFileRepository and supporting code to its own module
- Mark RocksDBFlowFileRepository as deprecated, both in code and documentation
- Log deprecation warning at startup if RocksDBFlowFileRepository is used
- Move native RocksDB info logs to NiFi debug level logs


# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ ] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

kevdoran commented:
To test this, I configured this in nifi.properties and tested a simple flow is still working as expected:

```
nifi.flowfile.repository.implementation=org.apache.nifi.controller.repository.RocksDBFlowFileRepository
```

This PR changes very little behavior, it is mostly just maven project restructuring.

exceptionfactory reviewed (commented):
Thanks for making these adjustments and upgrading the dependency @kevdoran!

Reviewing the dependency changes, the `rocksdbjni` has grown from 29 MB to 52 MB between version 6.6.4 and 6.29.5.

What do you think about taking the modularization one step further and separating it into a new NAR file with an optional build profile?

joewitt commented:
ahh good catch.  We should flag this in the migration guidance too that this module is no longer bundled by default but is available like other nars.  We should flag that in the docs whereever this thing is referenced too

kevdoran commented:
Thanks for the review and feedback @exceptionfactory! This is a good idea, and given the limited usage of this implementation, excluding it from the convenience binary makes sense. I've moved the implementation to its own NAR now and here are the new testing instructions:

```
# Build full project

# Copy the new nifi-rocksdb-nar to the assembly lib dir:
> cp nifi-nar-bundles/nifi-rocksdb-bundle/nifi-rocksdb-nar/target/nifi-rocksdb-nar-1.17.0-SNAPSHOT.nar nifi-assembly/target/nifi-1.17.0-SNAPSHOT-bin/nifi-1.17.0-SNAPSHOT/lib

# Edit the assembly nifi.properties to have the following: nifi.flowfile.repository.implementation=org.apache.nifi.controller.repository.RocksDBFlowFileRepository

# Start NiFi and verify it works as expected
> ./nifi-assembly/target/nifi-1.17.0-SNAPSHOT-bin/nifi-1.17.0-SNAPSHOT/bin/nifi.sh run
```

This is still failing at runtime when using those steps, so I need to fix something 🙂

@joewitt - This PR will update the Administration Guide appropriately, and I've added a note to the JIRA about the need for special migration guidance

exceptionfactory commented on nifi-nar-bundles/nifi-rocksdb-bundle/nifi-rocksdb-nar/pom.xml:
```diff
@@ -0,0 +1,41 @@
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
```
It looks like this NAR dependency tree is missing `nifi-rocksdb-repository`.

kevdoran replied:
Good catch, corrected in the latest

exceptionfactory reviewed (approved):
Thanks for making the adjustments @kevdoran! Verified successful configuration of the RocksDB repository after adding the `nifi-rocksdb-nar` to the `lib` directory. +1 merging
